File: go/tokenprovider.go
package keycloak

import (
	"context"
	"sync"
	"time"

	"golang.org/x/sync/singleflight"
)

// TokenProvider supplies access tokens to the admin facade — the only glue
// between auth and admin. Consumers may inject a custom implementation.
type TokenProvider interface {
	Token(ctx context.Context) (string, error)
}

// TokenSource obtains a fresh token set (e.g. via client-credentials).
type TokenSource func(ctx context.Context) (*TokenSet, error)

type clientCredentialsProvider struct {
	src     TokenSource
	skewSec int64
	group   singleflight.Group

	mu       sync.Mutex
	token    string
	expireAt int64 // epoch sec, skew-adjusted
}

// NewClientCredentialsTokenProvider caches a token and refreshes it before
// expiry, collapsing concurrent refreshes via single-flight.
func NewClientCredentialsTokenProvider(src TokenSource, skewSec int64) TokenProvider {
	return &clientCredentialsProvider{src: src, skewSec: skewSec}
}

func (p *clientCredentialsProvider) Token(ctx context.Context) (string, error) {
	p.mu.Lock()
	if p.token != "" && time.Now().Unix() < p.expireAt {
		tok := p.token
		p.mu.Unlock()
		return tok, nil
	}
	p.mu.Unlock()

	v, err, _ := p.group.Do("token", func() (any, error) {
		ts, err := p.src(ctx)
		if err != nil {
			return nil, err
		}
		p.mu.Lock()
		p.token = ts.AccessToken
		p.expireAt = time.Now().Unix() + max64(0, ts.ExpiresIn-p.skewSec)
		p.mu.Unlock()
		return ts.AccessToken, nil
	})
	if err != nil {
		return "", err
	}
	return v.(string), nil
}

func max64(a, b int64) int64 {
	if a > b {
		return a
	}
	return b
}

File: node/src/token-provider.ts
import type { TokenSet } from './tokens.js'

/** admin이 소비하는 액세스 토큰의 공급 인터페이스(auth와 admin을 잇는 유일한 접착제). */
export interface TokenProvider {
  getAccessToken(): Promise<string>
}

/** client-credentials 토큰을 발급할 수 있는 소스(AuthClient가 구현). */
export interface TokenSource {
  clientCredentialsToken(): Promise<TokenSet>
}

/**
 * client-credentials로 토큰을 자동 획득·캐시·갱신하는 기본 TokenProvider.
 * 만료 전이면 캐시를 재사용하고, 동시 요청은 단일 발급(single-flight)으로 합친다.
 */
export class ClientCredentialsTokenProvider implements TokenProvider {
  #cached?: { token: string; expiresAt: number }
  #inflight?: Promise<string>

  constructor(
    private readonly source: TokenSource,
    private readonly skewSeconds = 30,
  ) {}

  async getAccessToken(): Promise<string> {
    const cached = this.#cached
    if (cached && Date.now() < cached.expiresAt) return cached.token
    if (this.#inflight) return this.#inflight
    this.#inflight = this.#fetch()
    try {
      return await this.#inflight
    } finally {
      this.#inflight = undefined
    }
  }

  async #fetch(): Promise<string> {
    const ts = await this.source.clientCredentialsToken()
    this.#cached = {
      token: ts.accessToken,
      expiresAt: Date.now() + Math.max(0, ts.expiresIn - this.skewSeconds) * 1000,
    }
    return ts.accessToken
  }
}
